jormungandr/src/explorer/graphql/mod.rs

```diff
@@ -162,28 +202,51 @@ impl From<blockcfg::BlockDate> for BlockDate {
     }
 }
 
+#[derive(Clone)]
 struct Transaction {
     id: FragmentId,
-    in_block: HeaderHash,
+    block_hash: Option<HeaderHash>,
 }
 
 impl Transaction {
     fn from_id(id: FragmentId, context: &Context) -> FieldResult<Transaction> {
-        let in_block =
-            context
-                .db
-                .find_block_by_transaction(&id)
-                .wait()?
-                .ok_or(ErrorKind::NotFound(format!(
-                    "transaction not found: {}",
-                    &id,
-                )))?;
+        let block_hash = context
+            .db
+            .find_block_hash_by_transaction(&id)
+            .wait()
+            .unwrap()
+            .ok_or(ErrorKind::NotFound(format!(
+                "transaction not found: {}",
+                &id,
+            )))?;
+
+        Ok(Transaction {
+            id,
+            block_hash: Some(block_hash),
+        })
+    }
 
-        Ok(Transaction { id, in_block })
+    fn from_valid_id(id: FragmentId) -> Transaction {
+        Transaction {
+            id,
+            block_hash: None,
+        }
     }
 
     fn get_block(&self, context: &Context) -> FieldResult<ExplorerBlock> {
-        context.db.get_block(&self.in_block).wait()?.ok_or(
+        let block_id = match self.block_hash {
+            Some(block_id) => block_id,
+            None => context
+                .db
+                .find_block_hash_by_transaction(&self.id)
+                .wait()
+                .unwrap()
+                .ok_or(ErrorKind::InternalError(
+                    "Transaction's block was not found".to_owned(),
+                ))?,
+        };
+
+        context.db.get_block(&block_id).wait()?.ok_or(
```

Comment:
OK, though I suspect this dynamism is still not warranted and you can instead have two types for a transaction before its block is found and a transaction with a confirmed block, and transition from the first to the last as needed with a `self`-consuming method.

Reply:
I'll think about this, it's definitely interesting. The resolver must be implemented for just one type, but I'm thinking I can wrap an enum or something like that, like (just an on the fly example):

```rust
enum Transaction {
    UnconfirmedBlock { id: FragmentId },
    ConfirmedBlock { id: FragmentId, block: Block }
} 

impl Transaction {
    fn transition_method(self) -> Transaction { ... } 
}

#[juniper::object(
    Context = Context
)]
impl Transaction { resolvers here... }
```
Or maybe something using generics...

Reply:
The enum is not much an improvement over having an `Option<BlockHash>` as a member: same dynamism mostly, except now you also need to use an accessor method to get the reference to the invariable part for methods that don't have anything to do with the block hash.

Generics would get you into boxed objects and dynamic dispatch, as you still need to be able to hold values of both types in one API slot I guess? If so, don't bother.